File: src/color.c
#include "color.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void wsk_cairo_set_source_u32(cairo_t *cr, uint32_t color) {
    cairo_set_source_rgba(cr, (color >> (3 * 8) & 0xFF) / 255.0,
                          (color >> (2 * 8) & 0xFF) / 255.0,
                          (color >> (1 * 8) & 0xFF) / 255.0,
                          (color >> (0 * 8) & 0xFF) / 255.0);
}

uint32_t wsk_color_parse(const char *text, uint32_t fallback) {
    if (text[0] == '#') {
        ++text;
    }

    int len = strlen(text);
    if (len != 6 && len != 8) {
        fprintf(stderr, "Invalid color %s, defaulting to color 0x%08X\n", text,
                fallback);
        return fallback;
    }

    uint32_t color = (uint32_t) strtoul(text, NULL, 16);
    if (len == 6) {
        color = (color << 8) | 0xFF;
    }
    return color;
}

File: src/color.zig
const std = @import("std");

pub fn setSourceU32(cairo_context: ?*anyopaque, color: u32) void {
    const r = @as(f64, @floatFromInt((color >> 24) & 0xFF)) / 255.0;
    const g = @as(f64, @floatFromInt((color >> 16) & 0xFF)) / 255.0;
    const b = @as(f64, @floatFromInt((color >> 8) & 0xFF)) / 255.0;
    const a = @as(f64, @floatFromInt(color & 0xFF)) / 255.0;
    if (cairo_context) |p| {
        cairo_set_source_rgba(p, r, g, b, a);
    }
}

pub fn parse(text: []const u8, fallback: u32) u32 {
    var trimmed = text;
    if (trimmed.len > 0 and trimmed[0] == '#') {
        trimmed = trimmed[1..];
    }

    if (trimmed.len != 6 and trimmed.len != 8) {
        std.log.warn("Invalid color '{s}', defaulting to 0x{X:0>8}", .{ trimmed, fallback });
        return fallback;
    }

    const color_val = std.fmt.parseInt(u32, trimmed, 16) catch {
        std.log.warn("Invalid color '{s}', defaulting to 0x{X:0>8}", .{ trimmed, fallback });
        return fallback;
    };

    if (trimmed.len == 6) {
        return (color_val << 8) | 0xFF;
    }
    return color_val;
}

extern "c" fn cairo_set_source_rgba(cr: *anyopaque, r: f64, g: f64, b: f64, a: f64) void;
